crates/matrix-sdk-crypto/src/machine/mod.rs

```diff
@@ -1786,52 +1786,62 @@ impl OlmMachine {
-            // otherwise the sender_data would have been SenderVerified
-            let VerificationState::Unverified(verification_level) =
-                &encryption_info.verification_state
-            else {
-                unreachable!("inconsistent verification state");
-            };
-            Err(MegolmError::SenderIdentityNotTrusted(verification_level.clone()))
-        }
+        trace!(
+            verification_state = ?encryption_info.verification_state,
+            ?trust_requirement, "check_sender_trust_requirement",
+        );
 
-        match trust_requirement {
-            TrustRequirement::Untrusted => Ok(()),
-
-            TrustRequirement::CrossSignedOrLegacy => match &session.sender_data {
-                // Reject if the sender was previously verified, but changed
-                // their identity and is not verified any more.
-                SenderData::VerificationViolation(..) => Err(
-                    MegolmError::SenderIdentityNotTrusted(VerificationLevel::VerificationViolation),
-                ),
-                SenderData::SenderUnverified(..) => Ok(()),
-                SenderData::SenderVerified(..) => Ok(()),
-                SenderData::DeviceInfo { legacy_session: true, .. } => Ok(()),
-                SenderData::UnknownDevice { legacy_session: true, .. } => Ok(()),
-                _ => encryption_info_to_error(encryption_info),
-            },
+        // VerificationState::Verified is acceptable for all TrustRequirement levels, so
+        // let's get that out of the way
+        let verification_level = match &encryption_info.verification_state {
+            VerificationState::Verified => return Ok(()),
+            VerificationState::Unverified(verification_level) => verification_level,
+        };
+
+        let ok = match trust_requirement {
+            TrustRequirement::Untrusted => true,
+
+            TrustRequirement::CrossSignedOrLegacy => {
+                // `VerificationLevel::UnsignedDevice` and `VerificationLevel::None` correspond
+                // to `SenderData::DeviceInfo` and `SenderData::UnknownDevice`
+                // respectively, and those cases may be acceptable if the reason
+                // for the lack of data is that the sessions were established
+                // before we started collecting SenderData.
+                let legacy_session = match &session.sender_data {
+                    SenderData::DeviceInfo { legacy_session, .. } => legacy_session,
+                    SenderData::UnknownDevice { legacy_session, .. } => legacy_session,
+                    _ => &false,
+                };
+
+                match (verification_level, legacy_session) {
+                    (VerificationLevel::UnverifiedIdentity, _) => true,
+                    (VerificationLevel::UnsignedDevice, true) => true,
+                    (VerificationLevel::None(_), true) => true,
+                    _ => false,
```

Comment:
correspondances between `VerificationLevel` and `SenderData`:

`SenderData::VerificationViolation` -> `VerificationLevel::VerificationViolation`
`SenderData::SenderUnverified` -> `VerificationLevel::UnverifiedIdentity`
`SenderData::DeviceInfo` ->  `VerificationLevel::UnsignedDevice`
`SenderData::UnknownDevice` -> `VerificationLevel::None`

There is no equivalent to `SenderData::SenderVerified`; that is handled as `VerificationState::Verified` above.